gateway/join.go

```diff
@@ -0,0 +1,263 @@
+	}
+	txPkt.payload = cPayload
+
+	// send on rx2 window - opens 6 seconds after join request.
+	if !utils.SelectContextOrWait(ctx, time.Second*joinRx2WindowSec) {
+		return nil
+	}
+
+	// lock so there is not two sends at the same time.
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	errCode := int(C.send(&txPkt))
+	if errCode != 0 {
+		return errors.New("failed to send join accept packet")
+	}
+
+	return nil
+}
+
+// payload of join request consists of
+// | MHDR | JOIN EUI | DEV EUI  |   DEV NONCE  | MIC   |
+// | 1 B  |   8 B    |    8 B   |     2 B      |  4 B  |
+// https://lora-alliance.org/wp-content/uploads/2020/11/lorawan1.0.3.pdf page 34 for more info on join request.
+func (g *Gateway) parseJoinRequestPacket(payload []byte) (joinRequest, *Device, error) {
+	var joinRequest joinRequest
+
+	// everything in the join request payload is little endian
+	joinRequest.joinEUI = payload[1:9]
+	joinRequest.devEUI = payload[9:17]
+	joinRequest.devNonce = payload[17:19]
+	joinRequest.mic = payload[19:23]
+
+	matched := &Device{}
+
+	// device.devEUI is in big endian - reverse to compare and find device.
+	devEUIBE := reverseByteArray(joinRequest.devEUI)
+
+	// match the dev eui to gateway device
+	for _, device := range g.devices {
+		if bytes.Equal(device.devEui, devEUIBE) {
+			matched = device
+		}
+	}
+
+	if matched.name == "" {
+		g.logger.Debugf("received join requested with dev EUI %x - unknown device, ignoring", devEUIBE)
+		return joinRequest, nil, errNoDevice
+	}
+
+	err := validateMIC(matched.AppKey, payload)
+	if err != nil {
+		return joinRequest, nil, err
+	}
+
+	return joinRequest, matched, nil
+
+}
+
+// Format of Join Accept message:
+// | MHDR | JOIN NONCE | NETID |   DEV ADDR  | DL | RX DELAY |   CFLIST   | MIC  |
+// | 1 B  |     3 B    |   3 B |     4 B     | 1B |    1B    |  0 or 16   | 4 B  |
+// https://lora-alliance.org/wp-content/uploads/2020/11/lorawan1.0.3.pdf page 35 for more info on join accept.
+func (d *Device) generateJoinAccept(ctx context.Context, jr joinRequest) ([]byte, error) {
+	// generate random join nonce.
+	jn := generateJoinNonce()
+
+	// generate a random device address to identify uplinks.
+	d.addr = generateDevAddr()
+
+	// the join accept payload needs everything to be LE, so reverse the BE fields.
+	netIDLE := reverseByteArray(netID)
+	jnLE := reverseByteArray(jn)
+	dAddrLE := reverseByteArray(d.addr)
+
+	payload := make([]byte, 0)
+	payload = append(payload, 0x20)
+	payload = append(payload, jnLE[:]...)
+	payload = append(payload, netIDLE[:]...)
+	payload = append(payload, dAddrLE[:]...)
+	payload = append(payload, 0x00) // dl settings: default
+	payload = append(payload, 0x01) // rx delay: 1 second
+
+	// generate MIC
+	resMIC, err := crypto.ComputeLegacyJoinAcceptMIC(d.AppKey, payload)
+	if err != nil {
+		return nil, err
+	}
+
+	// everything but the mhdr needs to be encrypted.
+	payload = payload[1:]
+
+	payload = append(payload, resMIC[:]...)
+
+	enc, err := crypto.EncryptJoinAccept(d.AppKey, payload)
```

Comment:
Do we want to completely swallow the error? I'd log it if so, even if Debug, or use errors.Join to append multiple errors together if we should continue even in an error state.

Reply:
oops no that was not intentional, we should return the error here. fixing